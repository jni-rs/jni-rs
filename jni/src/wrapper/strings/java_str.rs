use std::os::raw::c_char;

use std::borrow::Cow;

use JNIEnv;

use objects::JString;

use errors::*;

use strings::JNIStr;

/// Reference to a string in the JVM. Holds a pointer to the array
/// returned by GetStringUTFChars. Calls ReleaseStringUTFChars on Drop.
/// Can be converted to a `&JNIStr` with the same cost as the `&CStr.from_ptr`
/// conversion.
pub struct JavaStr<'a> {
    internal: *const c_char,
    obj: JString<'a>,
    env: &'a JNIEnv<'a>,
}

impl<'a> JavaStr<'a> {
    /// Build a `JavaStr` from an object and a reference to the environment. You
    /// probably want to use `JNIEnv::get_string` instead.
    pub fn from_env(env: &'a JNIEnv<'a>, obj: JString<'a>) -> Result<Self> {
        let ptr = unsafe { env.get_string_utf_chars(obj)? };
        let java_str = JavaStr {
            internal: ptr,
            env: env,
            obj: obj,
        };
        Ok(java_str)
    }

    /// Extract the raw C string pointer from the JavaStr. This will be
    /// encoded using the JVM internal `CESU-8`-style.
    pub fn get_raw(&self) -> *const c_char {
        self.internal
    }
}

impl<'a> ::std::ops::Deref for JavaStr<'a> {
    type Target = JNIStr;
    fn deref(&self) -> &Self::Target {
        self.into()
    }
}

impl<'a> From<&'a JavaStr<'a>> for &'a JNIStr {
    fn from(other: &'a JavaStr) -> &'a JNIStr {
        unsafe { JNIStr::from_ptr(other.internal) }
    }
}

impl<'a> From<&'a JavaStr<'a>> for Cow<'a, str> {
    fn from(other: &'a JavaStr) -> Cow<'a, str> {
        let jni_str: &JNIStr = &*other;
        jni_str.into()
    }
}

impl<'a> From<JavaStr<'a>> for String {
    fn from(other: JavaStr) -> String {
        let cow: Cow<str> = (&other).into();
        cow.into_owned()
    }
}

impl<'a> Drop for JavaStr<'a> {
    fn drop(&mut self) {
        match unsafe { self.env.release_string_utf_chars(self.obj, self.internal) } {
            Ok(()) => {}
            Err(e) => warn!("error dropping java str: {}", e),
        }
    }
}
