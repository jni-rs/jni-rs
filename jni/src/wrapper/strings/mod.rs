// String types for sending to/from the jvm
mod ffi_str;
pub use self::ffi_str::*;

mod java_str;
pub use self::java_str::*;
