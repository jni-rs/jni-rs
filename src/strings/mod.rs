// String types for sending to/from the jvm
mod ffi_str;
pub use self::ffi_str::*;

mod mutf8_chars;
pub use self::mutf8_chars::*;

pub use crate::jvalue::{char_from_java, char_from_java_int, char_to_java, char_to_java_int};
