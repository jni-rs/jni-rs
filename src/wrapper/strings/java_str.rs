use std::{borrow::Cow, os::raw::c_char};

use log::warn;

use crate::{errors::*, objects::JString, strings::JNIStr, JNIEnv};

/// Reference to a string in the JVM. Holds a pointer to the array
/// returned by GetStringUTFChars. Calls ReleaseStringUTFChars on Drop.
/// Can be converted to a `&JNIStr` with the same cost as the `&CStr.from_ptr`
/// conversion.
pub struct JavaStr<'a: 'b, 'b> {
    internal: *const c_char,
    obj: JString<'a>,
    env: &'b JNIEnv<'a>,
}

impl<'a: 'b, 'b> JavaStr<'a, 'b> {
    /// Build a `JavaStr` from an object and a reference to the environment. You
    /// probably want to use `JNIEnv::get_string` instead.
    pub fn from_env(env: &'b JNIEnv<'a>, obj: JString<'a>) -> Result<Self> {
        let ptr = env.get_string_utf_chars(obj)?;
        let java_str = JavaStr {
            internal: ptr,
            env,
            obj,
        };
        Ok(java_str)
    }

    /// Extract the raw C string pointer from the JavaStr. This will be
    /// encoded using the JVM internal `CESU-8`-style.
    pub fn get_raw(&self) -> *const c_char {
        self.internal
    }
}

impl<'a: 'b, 'b> ::std::ops::Deref for JavaStr<'a, 'b> {
    type Target = JNIStr;
    fn deref(&self) -> &Self::Target {
        self.into()
    }
}

impl<'a: 'b, 'b: 'c, 'c> From<&'c JavaStr<'a, 'b>> for &'c JNIStr {
    fn from(other: &'c JavaStr) -> &'c JNIStr {
        unsafe { JNIStr::from_ptr(other.internal) }
    }
}

impl<'a: 'b, 'b: 'c, 'c> From<&'c JavaStr<'a, 'b>> for Cow<'c, str> {
    fn from(other: &'c JavaStr) -> Cow<'c, str> {
        let jni_str: &JNIStr = &*other;
        jni_str.into()
    }
}

impl<'a: 'b, 'b> From<JavaStr<'a, 'b>> for String {
    fn from(other: JavaStr) -> String {
        let cow: Cow<str> = (&other).into();
        cow.into_owned()
    }
}

impl<'a: 'b, 'b> Drop for JavaStr<'a, 'b> {
    fn drop(&mut self) {
        match self.env.release_string_utf_chars(self.obj, self.internal) {
            Ok(()) => {}
            Err(e) => warn!("error dropping java str: {}", e),
        }
    }
}
