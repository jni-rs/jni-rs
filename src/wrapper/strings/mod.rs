// String types for sending to/from the jvm
mod ffi_str;
pub use self::ffi_str::*;

mod mutf8_chars;
pub use self::mutf8_chars::*;
