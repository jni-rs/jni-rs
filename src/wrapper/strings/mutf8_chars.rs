use jni_sys::{jboolean, JNI_TRUE};
use std::{borrow::Cow, marker::PhantomData, os::raw::c_char};

use log::warn;

use crate::{
    env::JNIEnv,
    errors::*,
    objects::{JObjectRef, JString},
    strings::{JNIStr, JNIString},
    JavaVM,
};

/// Borrows the contents of a `java.lang.String` object, in Java's [modified
/// UTF-8] encoding.
///
/// This guard type is returned by [JString::mutf8_chars] and represents the
/// borrowed contents of a `java.lang.String` object that will be automatically
/// released when dropped.
///
/// This can be dereferenced to obtain a [`JNIStr`] which can in turn be
/// converted to a utf8 Rust string. (See [`JNIStr::to_str`] or [`JNIStr::to_string`]).
///
/// For example:
///
/// ```
/// # use jni::{errors::Result, env::JNIEnv, objects::*};
/// #
/// # fn f(env: &mut JNIEnv) -> Result<()> {
/// let string = env.new_string(c"Hello, world!")?;
/// let rust_utf8_string = string.mutf8_chars(env)?.to_string();
/// # Ok(())
/// # }
/// ```
///
/// # JNI String Types
///
/// From the point of view of JNI a [JString] is merely a reference to a
/// `java.lang.String` object and to access the underlying data, you need to use
/// JNI ([JString::borrow_utf_chars]) to explicitly borrow the underlying bytes
/// of the string.
///
/// [JNIStr] is to [JNIString] as `str` is to `String` or `CStr` is to
/// `CString`.
///
/// [JNIStr] and [JNIString] represent nul-terminated strings, like [`CStr`] and
/// [`CString`], that are encoded in [modified UTF-8].
///
/// This type is a guard that holds a temporary [JNIStr] reference to the
/// underlying bytes of a `java.lang.String` object.
///
/// [modified UTF-8]: https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8
pub struct MUTF8Chars<'local, StringRef>
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    obj: StringRef,
    chars: *const c_char,
    is_copy: bool,
    _lifetime: PhantomData<&'local ()>,
}

/// Borrows the contents of a `java.lang.String` object, in Java's [modified
/// UTF-8] encoding.
#[deprecated(note = "Renamed to MUTF8Chars, use JString::mutf8_chars() to get it")]
pub type JavaStr<'local, StringRef> = MUTF8Chars<'local, StringRef>;

impl<'local, StringRef> MUTF8Chars<'local, StringRef>
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    /// Constructs a [`MUTF8Chars`] from a `JNIEnv` and a `JString`.
    ///
    /// The string will be `NULL` terminated and encoded as [Modified
    /// UTF-8](https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8) /
    /// [CESU-8](https://en.wikipedia.org/wiki/CESU-8).
    ///
    /// The implementation may either create a copy of the character array for
    /// the given `String` or it may pin it to avoid it being collected by the
    /// garbage collector.
    ///
    /// Returns a [`MUTF8Chars`] that will automatically release the underlying
    /// character array when dropped (see [Self::release_string_utf_chars]).
    pub(crate) fn from_get_string_utf_chars(env: &JNIEnv<'_>, obj: StringRef) -> Result<Self> {
        let obj = null_check!(obj, "get_string_utf_chars obj argument")?;

        // SAFETY:
        // - We have checked that the object is not null.
        // - Having a `JString` guarantees that the reference is for a `java.lang.String`
        //   (it would require unsafe code for that to be violated)
        // - The pointer is immediately wrapped to ensure that the pointer will
        //   be released when dropped.
        unsafe {
            let mut is_copy: jboolean = false;
            let ptr: *const c_char = jni_call_only_check_null_ret!(
                env,
                v1_1,
                GetStringUTFChars,
                obj.as_raw(),
                &mut is_copy as *mut _
            )?;

            let is_copy = is_copy == JNI_TRUE;
            Ok(Self {
                obj,
                chars: ptr,
                is_copy,
                _lifetime: PhantomData,
            })
        }
    }

    /// Destroys the [`MUTF8Chars`] without freeing the underlying contents, and
    /// returns a raw pointer to it.
    ///
    /// The returned pointer is the same as the one returned by the
    /// [`as_ptr`][JNIStr::as_ptr] method. It points to a null-terminated
    /// string in [modified UTF-8] encoding (which is similar, but not
    /// identical, to [CESU-8]). It is valid when returned by this method, and
    /// will remain valid until freed (see below).
    ///
    /// [CESU-8]: https://en.wikipedia.org/wiki/CESU-8
    /// [modified UTF-8]: https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8
    ///
    /// This also returns the status for whether the original string was copied or not.
    ///
    /// # Warning
    ///
    /// After calling this method, the underlying string must be manually
    /// freed. This can be done either by reconstructing the [`MUTF8Chars`] using
    /// [`MUTF8Chars::from_raw`] and then dropping it, or by passing the pointer
    /// to the JNI function `ReleaseStringUTFChars`.
    pub fn into_raw(self) -> (*const c_char, bool) {
        let _dont_call_drop = std::mem::ManuallyDrop::new(self);
        (_dont_call_drop.chars, _dont_call_drop.is_copy)
    }

    /// Constructs a [`MUTF8Chars`] from raw components.
    ///
    /// The required components are, a [`JNIEnv`], a reference to a
    /// `java.lang.String` object, and a pointer to the characters of that
    /// `java.lang.String`.
    ///
    /// # Safety
    ///
    /// `ptr` must be a valid, non-null pointer, previously returned by
    /// [`MUTF8Chars::into_raw`] or the JNI function `GetStringUTFChars`. `ptr`
    /// must not belong to another [`MUTF8Chars`] at the same time.
    ///
    /// `obj` must be a non-null reference to the same `java.lang.String` object
    /// that `ptr` was obtained from.
    ///
    /// `is_copy` must be a boolean indicating whether the string was copied or
    /// not (as returned by `GetStringUTFChars`).
    ///
    /// # Example
    /// ```rust,no_run
    /// # use jni::{errors::Result, env::JNIEnv, strings::MUTF8Chars};
    /// #
    /// # fn example(env: &mut JNIEnv) -> Result<()> {
    /// let jstring = env.new_string(c"foo")?;
    /// let java_str = env.get_string(&jstring)?;
    ///
    /// let (ptr, is_copy) = java_str.into_raw();
    /// // Do whatever you need with the pointer
    /// let java_str = unsafe { MUTF8Chars::from_raw(env, &jstring, ptr, is_copy) };
    /// # Ok(())
    /// # }
    /// ```
    pub unsafe fn from_raw(
        _env: &JNIEnv<'_>,
        obj: StringRef,
        ptr: *const c_char,
        is_copy: bool,
    ) -> Self {
        Self {
            obj,
            chars: ptr,
            is_copy,
            _lifetime: PhantomData,
        }
    }

    /// Returns whether the string was copied or not.
    pub fn is_copy(&self) -> bool {
        self.is_copy
    }
}

impl<'local, StringRef> ::std::ops::Deref for MUTF8Chars<'local, StringRef>
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    type Target = JNIStr;
    fn deref(&self) -> &Self::Target {
        self.into()
    }
}

impl<'local, 'java_str, StringRef> From<&'java_str MUTF8Chars<'local, StringRef>>
    for &'java_str JNIStr
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    fn from(other: &'java_str MUTF8Chars<'local, StringRef>) -> &'java_str JNIStr {
        unsafe { JNIStr::from_ptr(other.chars) }
    }
}

impl<'local, StringRef> From<MUTF8Chars<'local, StringRef>> for JNIString
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    fn from(other: MUTF8Chars<'local, StringRef>) -> JNIString {
        let jni_str: &JNIStr = &other;
        jni_str.to_owned()
    }
}

impl<'local, 'java_str, StringRef> From<&'java_str MUTF8Chars<'local, StringRef>>
    for Cow<'java_str, str>
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    fn from(other: &'java_str MUTF8Chars<'local, StringRef>) -> Cow<'java_str, str> {
        let jni_str: &JNIStr = other;
        jni_str.into()
    }
}

impl<'local, StringRef> From<MUTF8Chars<'local, StringRef>> for String
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    fn from(other: MUTF8Chars<'local, StringRef>) -> String {
        let cow: Cow<str> = (&other).into();
        cow.into_owned()
    }
}

impl<'local, StringRef> Drop for MUTF8Chars<'local, StringRef>
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    fn drop(&mut self) {
        unsafe fn release_string_utf_chars(
            obj: jni_sys::jobject,
            chars: *const c_char,
        ) -> Result<()> {
            // Panic: Since we can't construct a `MUTF8Chars` without a valid `JNIEnv` reference we know
            // `JavaVM::singleton()` must be initialized and won't panic.
            JavaVM::singleton()?.with_env_current_frame(|env| {
                // This method is safe to call in case of pending exceptions (see the chapter 2 of the spec)
                jni_call_unchecked!(env, v1_1, ReleaseStringUTFChars, obj, chars);

                Ok(())
            })
        }

        match unsafe { release_string_utf_chars(self.obj.as_raw(), self.chars) } {
            Ok(()) => {}
            Err(e) => warn!("error dropping java str: {}", e),
        }
    }
}

impl<'local, StringRef> AsRef<JNIStr> for MUTF8Chars<'local, StringRef>
where
    StringRef: AsRef<JString<'local>> + JObjectRef,
{
    fn as_ref(&self) -> &JNIStr {
        self
    }
}
