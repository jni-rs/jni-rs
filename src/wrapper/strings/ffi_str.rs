use std::{
    borrow::{Borrow, Cow, ToOwned},
    ffi,
    os::raw::c_char,
};

use cesu8::{from_java_cesu8, to_java_cesu8};

use wrapper::strings::ffi_str;

/// Wrapper for `std::ffi::CString` that also takes care of encoding between
/// UTF-8 and Java's Modified UTF-8. As with `CString`, this implements `Deref`
/// to `&JNIStr`.
pub struct JNIString {
    internal: ffi::CString,
}

/// Wrapper for `std::ffi::CStr` that also takes care of encoding between
/// UTF-8 and Java's Modified UTF-8.
pub struct JNIStr {
    internal: ffi::CStr,
}

impl ::std::ops::Deref for JNIString {
    type Target = JNIStr;

    fn deref(&self) -> &Self::Target {
        unsafe { &*(self.internal.as_bytes_with_nul() as *const [u8] as *const ffi_str::JNIStr) }
    }
}

impl ::std::ops::Deref for JNIStr {
    type Target = ffi::CStr;

    fn deref(&self) -> &Self::Target {
        &self.internal
    }
}

impl<T> From<T> for JNIString
where
    T: AsRef<str>,
{
    fn from(other: T) -> Self {
        let enc = to_java_cesu8(other.as_ref()).into_owned();
        JNIString {
            internal: unsafe { ffi::CString::from_vec_unchecked(enc) },
        }
    }
}

impl<'a> From<&'a JNIStr> for Cow<'a, str> {
    fn from(other: &'a JNIStr) -> Cow<'a, str> {
        let bytes = other.to_bytes();
        match from_java_cesu8(bytes) {
            Ok(s) => s,
            Err(e) => {
                debug!("error decoding java cesu8: {:#?}", e);
                String::from_utf8_lossy(bytes)
            }
        }
    }
}

impl From<JNIString> for String {
    fn from(other: JNIString) -> String {
        Cow::from(other.borrowed()).into_owned()
    }
}

impl JNIString {
    /// Get the borrowed version of the JNIString. Equivalent to
    /// `CString::borrowed`.
    pub fn borrowed(&self) -> &JNIStr {
        self
    }
}

impl JNIStr {
    /// Construct a reference to a `JNIStr` from a pointer. Equivalent to
    /// `CStr::from_ptr`.
    pub unsafe fn from_ptr<'a>(ptr: *const c_char) -> &'a JNIStr {
        &*(ffi::CStr::from_ptr(ptr) as *const ffi::CStr as *const ffi_str::JNIStr)
    }
}

// impls for CoW
impl Borrow<JNIStr> for JNIString {
    fn borrow(&self) -> &JNIStr {
        self
    }
}

impl ToOwned for JNIStr {
    type Owned = JNIString;

    fn to_owned(&self) -> JNIString {
        unsafe {
            JNIString {
                internal: ffi::CString::from_vec_unchecked(self.to_bytes().to_vec()),
            }
        }
    }
}
