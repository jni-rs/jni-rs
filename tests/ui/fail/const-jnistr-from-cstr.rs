use jni::strings::JNIStr;

fn main() {
    const INVALID_MUTF8_CSTR: &JNIStr = JNIStr::from_cstr(c"invalid mutf8: ðŸ¦€");
}
