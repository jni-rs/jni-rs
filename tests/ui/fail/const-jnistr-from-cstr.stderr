error[E0080]: evaluation panicked: JNIStr::from_cstr: input is not valid modified UTF-8
 --> tests/ui/fail/const-jnistr-from-cstr.rs:4:41
  |
4 |     const INVALID_MUTF8_CSTR: &JNIStr = JNIStr::from_cstr(c"invalid mutf8: ðŸ¦€");
  |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `main::INVALID_MUTF8_CSTR` failed inside this call
  |
note: inside `JNIStr::from_cstr`
 --> src/strings/ffi_str.rs
  |
  |             panic!("JNIStr::from_cstr: input is not valid modified UTF-8");
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the failure occurred here
